### PR TITLE
fix: D2H copy of expanded tensor reads stride-0 dims as garbage

### LINKED ISSUE #1470: [BUG] : [Ministral-3-14B-Instruct-2512] expand() produces incorrect values — stride-0 dim materialized as raw memory copy
### What happened?

When calling .expand() on a tensor residing on the Spyre device and then transferring it back to CPU via .cpu(), the resulting tensor contains incorrect values and wrong strides.
The expanded tensor on-device correctly shows stride 0 on the expanded dimension — matching expected PyTorch behavior. However, after .cpu(), the strides change from (4, 1, 0) to (12, 3, 1), indicating the tensor was materialized as if contiguous during the device-to-host copy.
This causes the copy to read beyond the actual allocated storage (only 8 bytes exist for the original [1, 4, 1] tensor, but 24 bytes are read), producing garbage values in the output.
The same operation on CPU produces correct results. The bug is isolated to the Spyre→CPU transfer path for non-contiguous tensors created via expand().

**Example :**
```
src shape: [1, 4, 1]
Memory layout (8 bytes total, 4 values):
[ 1.00, -1.37,  1.39, -1.85 ]

after expand shape:[1,4,3]

Logical view [1, 4, 3] with strides (4, 1, 0):

        col0    col1    col2
row0  [  1.00,   1.00,   1.00  ]   ← all point to v0
row1  [ -1.37,  -1.37,  -1.37  ]   ← all point to v1
row2  [  1.39,   1.39,   1.39  ]   ← all point to v2
row3  [ -1.85,  -1.85,  -1.85  ]   ← all point to v3

Still only 8 bytes in memory. No new data allocated.

What we got as output with strides [12, 3, 1]:

Spyre expanded            :
 tensor([[ 1.0000e+00, -1.3691e+00,  1.3926e+00],
        [-1.8457e+00,  3.5059e+00,  2.5234e+01],
        [ 5.0625e+01,  1.7234e+01,  7.0953e-04],
        [-4.2531e+01,  1.0112e+02,  0.0000e+00]],

```

### What did you expect to happen?

When calling .expand() followed by .cpu() on a Spyre tensor, the resulting CPU tensor should contain the same values as the equivalent CPU operation.
Specifically,  each expanded column should repeat the original value, not read sequential memory. The expected output for a [1, 4, 1] tensor expanded to [1, 4, 3] is:
[  1.00,   1.00,   1.00 ]
[ -1.37,  -1.37,  -1.37 ]
[  1.39,   1.39,   1.39 ]
[ -1.85,  -1.85,  -1.85 ]

### How can we reproduce it?

Just run this code : 

```
import torch
import torch_spyre

DEVICE = "spyre"

src = torch.randn(1, 4, 1, dtype=torch.float16)

spyre_tensor = src.to(DEVICE).expand(1, 4, 3)
spyre_out    = spyre_tensor.cpu()


cpu_out = src.expand(1, 4, 3)


print("=== Strides ===")
print(f"Spyre strides (after .cpu()) : {spyre_out.stride()}")   
print(f"CPU   strides                : {cpu_out.stride()}")    

print("\n=== Values ===")
print(f"src            : {src.squeeze()}")
print(f"CPU  expanded  : {cpu_out.squeeze()}")
print(f"Spyre expanded : {spyre_out.squeeze()}")


print("\n=== Slice equality (expanded cols should all be equal) ===")
print(f"CPU   col0 == col1 : {torch.equal(cpu_out[:,:,0],   cpu_out[:,:,1])}")    
print(f"CPU   col0 == col2 : {torch.equal(cpu_out[:,:,0],   cpu_out[:,:,2])}")    
print(f"Spyre col0 == col1 : {torch.equal(spyre_out[:,:,0], spyre_out[:,:,1])}")  
print(f"Spyre col0 == col2 : {torch.equal(spyre_out[:,:,0], spyre_out[:,:,2])}")  

print("\n=== Tensor equality (cpu_out vs spyre_out) ===")
print(f"torch.equal : {torch.equal(cpu_out, spyre_out)}")                          
print(f"allclose    : {torch.allclose(cpu_out.float(), spyre_out.float())}")        
```

### Any environmental details we need to know?

Ensure the tests are executed within the torch-spyre environment.

### Anything else we need to know?

On-device behavior is correct:
The strides on Spyre before .cpu() are (4, 1, 0) — matching CPU exactly.
The bug only manifests during the transfer, not during the expand operation itself.

### Relevant log output

```Shell
(torch-spyre) [vishwasr@vishwasr-spyre-dev-pf model_centric_op_level_tests]$ python tets.py
=== Strides ===
Spyre strides (after .cpu()) : (12, 3, 1)
CPU   strides                : (4, 1, 0)

=== Values ===
src            : tensor([-1.2861,  0.6987, -1.7764,  1.7969], dtype=torch.float16)
CPU  expanded  : tensor([[-1.2861, -1.2861, -1.2861],
        [ 0.6987,  0.6987,  0.6987],
        [-1.7764, -1.7764, -1.7764],
        [ 1.7969,  1.7969,  1.7969]], dtype=torch.float16)
Spyre expanded : tensor([[-1.2852e+00,  6.9824e-01, -1.7754e+00],
        [ 1.7969e+00,  3.3600e+03,  3.2280e+03],
        [ 9.3312e+01,  0.0000e+00,  3.4304e+04],
        [ 3.0398e-06,  5.0562e+01,  5.0469e+01]], dtype=torch.float16)

=== Slice equality (expanded cols should all be equal) ===
CPU   col0 == col1 : True
CPU   col0 == col2 : True
Spyre col0 == col1 : False
Spyre col0 == col2 : False

=== Tensor equality (cpu_out vs spyre_out) ===
torch.equal : False
allclose    : False
```

### PR DESCRIPTION
**Pull request by AdnanHoque**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Fixes silent data corruption on D2H copy of non-contiguous Spyre tensors (e.g. `.expand()`/`.broadcast_to()` views). The DMA was sized using the underlying allocation, dropping stride-0 broadcast dims. Now materializes the broadcast CPU-side after a contiguous DMA.

#### Which issue(s) this PR is related to:

Fixes #1470

#### Special notes for your reviewer:

- The fix is scoped to the entry point (`spyre_copy_from`). The deeper
  bug lives in `generate_dci`'s D2H branch, where the
  dma_sizes/dma_strides assumption is baked in. The original code's comment at `spyre_mem.cpp:360`
  ("Transfer contiguous memory, deal with view on cpu") describes the
  same strategy this PR completes.
  
- A first attempt via `self.contiguous()` surfaced an unrelated
  deeptools scheduler crash (`L3DlOpsScheduler.cpp:1058` "There must be
  at least one valid candidate") for inner-stride-0 broadcasts.
  Repro: `torch.ops.spyre.copy_from_d2d(non_contig_src, contig_dst)`.
  Not in this PR's scope.

#### Does this PR introduce a user-facing change?

Yes — `.cpu()` on non-contiguous Spyre tensors now returns correct values.

### CODE REVIEW

File: torch_spyre/csrc/spyre_mem.cpp [hunk 1]
```diff
@@ -519,6 +519,38 @@ at::Tensor& spyre_set_storage(at::Tensor& result, at::Storage storage,
  */
 at::Tensor spyre_copy_from(const at::Tensor& self, const at::Tensor& dst,
                            bool non_blocking) {
+  // D2H of a broadcast/expand source (any dim with size>1 and stride==0): the
+  // DMA path uses dma_sizes/dma_strides and silently drops broadcast dims,
+  // leaving the CPU destination underfilled. Materialize on the CPU side: DMA
+  // the underlying allocation as a contiguous view, then realize self's
+  // logical sizes/strides on top of the staging buffer and copy into dst.
+  // Other non-contiguous layouts (transpose, column-major, sliced) are left to
+  // the existing path, which handles them correctly.
```
Comment (ani300) [inline torch_spyre/csrc/spyre_mem.cpp]: The contract for device-to-host copies in pytorch is that a tensor that is dense and non-overlapping will get their view respected on the copy, while tensors which are not that (the ones you're handling here, like expand) will get fully realized. Can you change your code to respect that? It would simply be calling contiguous() on the cpu tensor at the end. Also, if you can, use the new copyAsync API, as it's faster and helps deprecate sendnn quicker

Comment (ani300) [reply torch_spyre/csrc/spyre_mem.cpp]: Pytorch has a helper as well for checking if a tensor is dense and non-overlapping

Comment (AdnanHoque) [reply torch_spyre/csrc/spyre_mem.cpp]: yes, sounds good. switched to copyAsync API and added the .contiguous() to the CPU tensor to respect d2h contract

File: torch_spyre/csrc/spyre_mem.cpp [hunk 2]
```diff
@@ -552,6 +519,29 @@ at::Tensor& spyre_set_storage(at::Tensor& result, at::Storage storage,
  */
 at::Tensor spyre_copy_from(const at::Tensor& self, const at::Tensor& dst,
                            bool non_blocking) {
+  // The DMA path uses dma_sizes/dma_strides directly, so non-contiguous views
+  // (broadcasts, strided slices) drop dims and underfill or over-read dst.
+  // DMA the underlying allocation to a CPU staging buffer, lay self's logical
+  // view on top, and realize via contiguous() into dst.
+  if (self.is_privateuseone() && dst.is_cpu() &&
```
Comment (ani300) [inline torch_spyre/csrc/spyre_mem.cpp]: can this whole block be integrated into the line 546 conditional? Aka:

```cpp
  bool non_overlapping_and_dense = true;
  at::Tensor alloc_view, cpu_alloc, cpu_view;
  const at::Tensor* copy_from = &self, copy_to = &dst;
  if (dst.is_privateuseone()) {
    stream = getCurrentStream(dst.device());
    auto* spyre_impl = static_cast<SpyreTensorImpl*>(self.unsafeGetTensorImpl());
    if (!spyre_impl->is_non_overlapping_and_dense_default())) {
      non_overlapping_and_dense = false;
      c10::IntArrayRef alloc_sizes(spyre_impl->dma_sizes);
      c10::IntArrayRef alloc_strides(spyre_impl->dma_strides);
      alloc_view = at::as_strided(self, alloc_sizes, alloc_strides, /*storage_offset=*/0);
      cpu_alloc = at::empty(alloc_sizes, dst.options());
      copy_from = &alloc_view;
      copy_to = &cpu_alloc;
    }
  } else {
    stream = getCurrentStream(self.device());
  }
  stream.copyAsync(*copy_from, *copy_to);
  if (!non_blocking) {
    stream.synchronize();
  }
  if (!non_overlapping_and_dense) {
    at::Tensor cpu_view = cpu_alloc.as_strided(self.sizes(), self.strides(),
                                               self.storage_offset());
    dst.copy_(cpu_view);
  }
  return dst;
```

Comment (ani300) [reply torch_spyre/csrc/spyre_mem.cpp]: I think the contiguous() on the cpu view is not required as dst should have the right sizes and strides per the contract

Comment (AdnanHoque) [reply torch_spyre/csrc/spyre_mem.cpp]: Done, folded it into the same conditional and dropped contiguous()

**Review by ani300 — APPROVED** [review]
lgtm! thanks for your help with this

**Review by thoangtrvn — APPROVED** [review]
LGTM.  The PR first copy the compact data from device to CPU (a dedicated heap allocation); then use PyTorch's built-in mechanism to materialize that into the destination tensor.
